cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -0,0 +1,880 @@
+    int32_t* new_hash_slots =
+        reinterpret_cast<hash_slot_t*>(new_hash_table->mutable_data());
+    int64_t new_mod_bitmask = new_size - 1;
+
+    const uint8_t* dict_data = dict_data_.data();
+
+    for (int i = 0; i < hash_table_size_; ++i) {
+      hash_slot_t index = hash_slots_[i];
+
+      if (index == kHashSlotEmpty) {
+        continue;
+      }
+
+      // Find an empty slot in the new hash table
+      int64_t j = HashValue(dict_data + index * byte_width_) & new_mod_bitmask;
+      while (kHashSlotEmpty != new_hash_slots[j]) {
+        ++j;
+        if (ARROW_PREDICT_FALSE(j == hash_table_size_)) {
+          j = 0;
+        }
+      }
+
+      // Copy the old slot index to the new hash table
+      new_hash_slots[j] = index;
+    }
+
+    hash_table_ = new_hash_table;
+    hash_slots_ = reinterpret_cast<hash_slot_t*>(hash_table_->mutable_data());
+    hash_table_size_ = new_size;
+    mod_bitmask_ = new_size - 1;
+
+    return Status::OK();
+  }
+
+  int32_t byte_width_;
+  TypedBufferBuilder<uint8_t> dict_data_;
+  int32_t dict_size_;
+};
+
+// ----------------------------------------------------------------------
+// Unique implementation
+
+template <typename Type>
+class UniqueImpl : public HashTableKernel<Type, UniqueImpl<Type>> {
+ public:
+  static constexpr bool allow_expand = true;
+  using Base = HashTableKernel<Type, UniqueImpl<Type>>;
+  using Base::Base;
+
+  Status Reserve(const int64_t length) { return Status::OK(); }
+
+  void ObserveFound(const hash_slot_t slot) {}
+  void ObserveNull() {}
+  void ObserveNotFound(const hash_slot_t slot) {}
+
+  Status DoubleSize() { return Base::DoubleTableSize(); }
+
+  Status Append(const ArrayData& input) override { return Base::Append(input); }
+
+  Status Flush(std::vector<Datum>* out) override {
+    // No-op
+    return Status::OK();
+  }
+};
+
+// ----------------------------------------------------------------------
+// Dictionary encode implementation
+
+template <typename Type>
+class DictEncodeImpl : public HashTableKernel<Type, DictEncodeImpl<Type>> {
+ public:
+  static constexpr bool allow_expand = true;
+  using Base = HashTableKernel<Type, DictEncodeImpl>;
+
+  DictEncodeImpl(const std::shared_ptr<DataType>& type, MemoryPool* pool)
+      : Base(type, pool), indices_builder_(pool) {}
+
+  Status Reserve(const int64_t length) { return indices_builder_.Reserve(length); }
+
+  void ObserveNull() { indices_builder_.UnsafeAppendToBitmap(false); }
+
+  void ObserveFound(const hash_slot_t slot) { indices_builder_.UnsafeAppend(slot); }
+
+  void ObserveNotFound(const hash_slot_t slot) { return ObserveFound(slot); }
+
+  Status DoubleSize() { return Base::DoubleTableSize(); }
+
+  Status Flush(std::vector<Datum>* out) override {
+    std::shared_ptr<ArrayData> result;
+    RETURN_NOT_OK(indices_builder_.FinishInternal(&result));
+    out->push_back(Datum(result));
+    return Status::OK();
+  }
+
+  using Base::Append;
+
+ private:
+  Int32Builder indices_builder_;
```

Comment:
Why do we resort to an Int32Builder? Because at the end all chunks shall have the same integer type for their indices?

Reply:
Yes. In addition to being slower (more branching, possibility of appends failing), we would have to deal with promoting all chunks to the largest type at the end. I am open to using the adaptive builder but only after writing more benchmarks and doing some more analysis